### ISSUE TITLE
X13-ARIMA SEATS different results in statsmodels and R

### ISSUE THREAD

**Issue by as4456** [statsmodels/statsmodels#3549]
Hi,  
I am currently trying to get a seasonally adjusted series. I am getting different results using statsmodels (0.8.0) and R (seasonal package). 
I am using a Windows 7 64-bit machine and the version of X13 windows binary is winx13_V2.4.
I am attaching the input data file here.
Please see the brief script in Python and R below:

# Python

import pandas as pd
import statsmodels as sm
import os
import matplotlib.pyplot as plt

df_railroad=pd.read_excel(r'C:\Users\\Input_Data1.xlsx',sheetname="Sheet1",index_col=0,header=0,parse_cols="A:B")
df_railroad.columns=["Railroad"]

XPATH = os.chdir("C:\\Users\\winx13_V2.4\\winx13\\x13as")

temp_series = pd.DataFrame(df_railroad["Railroad"].values,index=df_railroad.index)
x13results = sm.tsa.x13.x13_arima_analysis(endog = temp_series,x12path=XPATH, outlier=True,print_stdout=True)

df_railroad_adj = pd.Series(x13results.seasadj.values,index=df_railroad.index.values)
df_railroad_adj=df_railroad_adj.to_frame()
df_railroad_adj.columns=["Adjusted"]
os.chdir(os.getcwd())

fig, ax = plt.subplots(figsize=(13,6))
ax.set_title("Actual and Seasonally Adjusted Series: US Railroad Traffic \n Monthly (Jan 2000 - Feb 2017)")
ax.plot(df_railroad_adj.index,df_railroad_adj["Adjusted"],linewidth=2, marker='', markersize=3, zorder=1,label="Seasonally Adjusted")
ax.plot(df_railroad_adj.index,df_railroad["Railroad"],linewidth=2, marker='', markersize=3, zorder=1,label="Actual")
plt.show()
[Input_Data1.xlsx](https://github.com/statsmodels/statsmodels/files/832703/Input_Data1.xlsx)

# R code
library(seasonal)
library(x13binary)
library(xlsx)
library(zoo)
library(xts)

checkX13()
df_rail=read.xlsx2("Input_Data1.xlsx",1,colClasses=c("Date","numeric"))
colnames(df_rail)=c("Date","Railroad")
df_rail<-na.omit(df_rail)
df_rail_ts<-ts(df_rail$Railroad,start=c(1990,01),frequency=12)
mod <- seas(df_rail_ts)
plot(mod)
plot(resid(mod), main = "Residuals")
qqnorm(resid(mod), main = "Residuals compared to Normal")
pacf(resid(mod), "Partial ACF of residuals")
df_out<-xts(mod$data, as.Date(df_rail$Date, format='%Y-%m-%d'))
write.zoo(df_out,"R_seasadj.csv",sep=",",col.names=TRUE)

**Comment by josef-pkt** [comment]
Can you summarize what the differences are?
Do you have the same data in the actual call?
you have `df_rail<-na.omit(df_rail)`, does this drop some nan rows in the R version?

**Comment by josef-pkt** [comment]
Unfortunately, none of the current developers is using x13. So maybe you get a faster answer on stackoverflow.
I'm neither familiar with the details of the options and their defaults in statsmodels, R nor X13 itself.

**Comment by as4456** [comment]
![python_plot](https://cloud.githubusercontent.com/assets/8718176/23777879/536a287c-0505-11e7-9a41-cacebb3f66e0.png)
![rplot](https://cloud.githubusercontent.com/assets/8718176/23777878/53681eb0-0505-11e7-9e3e-51ffd911b2f7.png)

**Comment by as4456** [comment]
The above plot is from Python and below from R

**Comment by josef-pkt** [comment]
one guess: maybe the freq is not recognized, The python plot doesn't seem to use seasonal correction.

Does the index in `temp_series` have monthly freq?
Try to specify freq=12 directly in the x13 call. 
http://www.statsmodels.org/devel/generated/statsmodels.tsa.x13.x13_arima_analysis.html

**Comment by josef-pkt** [comment]
You could also check if `x13results.stdout` shows any information about the correct interpretation of the data.

**Comment by as4456** [comment]
freq=12 gives the same results (It should anyways be provided if there is no date index).  
results.stdout does not provide any relevant information to identify the cause of difference.
I will look more into this, maybe changing arguments of maxorder

**Comment by as4456** [comment]
I will try to post this issue on stackoverflow.
One minor change in title of python plot title. It should mention "Monthly", incorrectly mentioned as "Weekly"

**Comment by josef-pkt** [comment]
detailed output is in results, I need to split by pages or only the last page shows in spyder

page zero shows where the temp log files are
print(x13results.results.split('\x0c')[0])

e.g.
```
>>> print(x13results.results.split('\x0c')[3])
        Unnamed Series            PAGE   3, SERIES Unnamed Series
                        Automatic ARIMA Model Selection

                    Procedure based closely on TRAMO method 
                         of Gomez and Maravall (2000)
              "Automatic Modeling Methods for Univariate Series",
                            A Course in Time Series
                  (Edited by D. Pena, G. C. Tiao, R. S. Tsay),
                         New York : J. Wiley and Sons


     Maximum order for regular ARMA parameters :   2
     Maximum order for seasonal ARMA parameters :   1
     Maximum order for regular differencing :   2
     Maximum order for seasonal differencing :   1

   Results of Unit Root Test for identifying orders of differencing:
     Regular difference order :   0
     Seasonal difference order :   1

     Mean is significant.

  Automatic model choice : (2 0 2)(0 1 1)

   Results of Unit Root Test for identifying orders of differencing:
     Regular difference order :   0
     Seasonal difference order :   1

     Mean is significant.

  Automatic model choice : (1 0 2)(0 1 1)


  Final Checks for Identified Model


  Checking for Unit Roots.
   No unit roots found.

  Checking for insignificant ARMA coefficients.

  Final automatic model choice : (3 0 1)(0 1 1)
   End of automatic model selection procedure.

 Average absolute percentage error in within-sample forecasts:
  Last year:  10.19      Last-1 year:  11.66     Last-2 year:   9.04
  Last three years:   10.30

 Estimation converged in   21 ARMA iterations,  163 function evaluations.

 Regression Model
 ------------------------------------------------------------------------------
                             Parameter        Standard
 Variable                     Estimate           Error      t-value
 ------------------------------------------------------------------------------
 Constant                       0.0168         0.00178         9.45

 Automatically Identified Outliers
   LS2009.Feb                  -0.1779         0.02607        -6.82
 ------------------------------------------------------------------------------


 ARIMA Model:  (3 0 1)(0 1 1)
   Seasonal differences:    1
                                              Standard
 Parameter                    Estimate          Errors
 -----------------------------------------------------
 Nonseasonal AR                                    
   Lag  1                       0.0766         0.05707
   Lag  2                       0.1809         0.04338
   Lag  3                       0.5954         0.04376

 Nonseasonal MA                                    
   Lag  1                       0.6225         0.06099

 Seasonal MA                                       
   Lag 12                       1.0000         0.04545

 Variance                       0.70496E-02
 SE of Var                      0.56262E-03
 -----------------------------------------------------

 Likelihood Statistics
 ------------------------------------------------------------------
 Number of observations (nobs)                                  326
 Effective number of observations (nefobs)                      314
 Number of parameters estimated (np)                              8
 Log likelihood                                            312.4382
 Transformation Adjustment                               -4578.0784
 Adjusted Log likelihood (L)                             -4265.6402
 AIC                                                      8547.2804 
 AICC (F-corrected-AIC)                                   8547.7525 
 Hannan Quinn                                             8559.2659 
 BIC                                                      8577.2756 
 ------------------------------------------------------------------

 QS Statistic for regARIMA Model Residuals (full series):             0.00
                                                       (P-Value =     1.0000)

 QS Statistic for regARIMA Model Residuals (starting 2009.M>> >ar):       0.00
                                                       (P-Value =     1.0000)


  At least one visually significant residual trading day peaks has been
  found in the spectral plot of the following series starting in 2009.Mar:

          regARIMA model residuals (1 Trading Day peak(s))




      10*LOG(SPECTRUM) of the regARIMA model residuals
      Spectrum estimated from 2009.Mar to 2017.Feb.
 ++++++++++I+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++I
     -13.83I                                                             I    -13.83
           I                                          T                  I
           I                                          T                  I
           I                                          T                  I
     -15.16I                                          T                  I    -15.16
           I                                          T                  I
           I                *          *              T                  I
           I                *          *              T                  I
     -16.50I                *         ***             T                  I    -16.50
           I                *         ***             T                  I
           I   *    *    *  *         ***             T                  I
           I   *    *    *  *         ***             T                  I
     -17.83I   *    *    *  *         ***             T                  I    -17.83
           I   *    *    *  *         ***             T                  I
           I   *    *    *  *         ***             T              *   I
           I   *    *    ** *         ***             T         T    *   I
     -19.16I   *    *    ** *         ***             T         T    *   I    -19.16
           I  **    *    ** *         ***             T         T    *   I
           I  **    *    ** *         ****            T         T    *   I
           I  **    *    ** *         ****            T        *T    *   I
     -20.49I  **    *    ****         ****            T        *T    *   I    -20.49
           I  **    *    *****        ****            T        *T    *   I
           I ***    *    *****        ****            T        *T    *   I
           I*****   *    *****        ****            T        *T    *   I
     -21.82I*****   *    *****       *****            T        *T    **  I    -21.82
           I*****   *    *****       *****            T        *T    **  I
           I*****   *    *****       *****            T        *T*   **  I
           I*****  **    *****       *****            T        *T*   **  I
     -23.16I*****  **    *****       *****            T        *T*  ***  I    -23.16
           I*****  **    *****       *****     *      T*       *T*  ***  I
           I*****  ***   *****       *****    ***     T*       *T*  ***  I
           I*****  ***   *****       *****S   ***     T*       *T*  ***  I
     -24.49I*****  ***   *****       *****S   ***     T*      S*T*  ***  I    -24.49
           I*****  ***  ******      ******S  ****     T*      S*T** ***  I
           I****** ***  ******      ******S  ****     T*      S*T******  I
           I****** ***  ******      ******S  ****     T*  **  S*T******  I
     -25.82I**********  ******      ******S  ****     T*  **  S*T******  I    -25.82
           I**********  ******     *******S  ****     T*  **  S*T******  I
           I**********  ******     *******S******     T*  **  S*T******* I
           I**********  ******     *******S*******   *T*  **  S*T******* I
     -27.15I**********  ******    ********S*******   *T*  ** *S*T******* I    -27.15
           I**********S *******   ********S*******   *T*  ****S*T******* I
           I**********S********   ********S*******   *T* *****S*T*******SI
           I**********S********   ********S*******   *T*******S*T*******SI
     -28.49I**********S********   ********S*******   *T*******S*T*******SI    -28.49
           I**********S********  *********S*******   *T*******S*T*******SI
           I**********S********  *********S*******   *T*******S*T*******SI
           I**********S********  *********S*******   *T*******S*T*******SI
     -29.82I**********S********  *********S********  *T*******S*T*******SI    -29.82
           I**********S*********S*********S******** S*T*******S*T*******SI
           I**********S*********S*********S******** S*T*******S*T*******SI
           I**********S*********S*********S******** S*T*******S*T*******SI
     -31.15I**********S*********S*********S*********S*T*******S*T*******SI    -31.15
 ++++++++++I+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++I
            S=SEASONAL FREQUENCIES, T=TRADING DAY FREQUENCIES


 FORECASTING
  Origin  2017.Feb
  Number        12

  Forecasts and Standard Errors of the Transformed Data
   ------------------------------
                         Standard
       Date   Forecast      Error
   ------------------------------
   2017.Mar      14.73      0.084
   2017.Apr      14.59      0.096
   2017.May      14.70      0.096
   2017.Jun      14.68      0.106
   2017.Jul      14.63      0.108
   2017.Aug      14.73      0.109
   2017.Sep      14.69      0.111
   2017.Oct      14.73      0.111
   2017.Nov      14.67      0.112
   2017.Dec      14.65      0.113
   2018.Jan      14.63      0.113
   2018.Feb      14.60      0.113
   ------------------------------

  Confidence intervals with coverage probability ( 0.95000
  On the Original Scale
   ------------------------------------------
       Date       Lower   Forecast      Upper
   ------------------------------------------
   2017.Mar  2114417.42 2494041.05 2941822.51
   2017.Apr  1802153.12 2174029.23 2622642.35
   2017.May  1998879.98 2414891.10 2917483.33
   2017.Jun  1925754.64 2368820.32 2913823.80
   2017.Jul  1820549.54 2248810.53 2777814.43
   2017.Aug  2009207.13 2485971.61 3075867.46
   2017.Sep  1926357.38 2394801.67 2977160.47
   2017.Oct  2007211.31 2496831.64 3105885.37
   2017.Nov  1881449.82 2343270.55 2918449.81
   2017.Dec  1855825.68 2314965.43 2887698.45
   2018.Jan  1802627.47 2248657.81 2805050.99
   2018.Feb  1753909.54 2189607.78 2733540.19
   ------------------------------------------
```

**Comment by josef-pkt** [comment]
I never looked at the x13 output.
What looks weird to me, if I interpret this correctly, is that seasonal ma `Lag 12  1.0000 0.04545` coefficient 1. might exactly compensate seasonal differencing (but I don't know which sign convention x13 uses)

**Comment by josef-pkt** [comment]
Trend looks good, but I'm not able to get anything seasonally adjusted that is has less fluctuation.
Also I don't seem to be able to use `maxorder` to limit the seasonal ma terms.

Maybe somebody has a better idea on stackoverflow

**Comment by josef-pkt** [comment]
seasonal_decompose doesn't have a problem detecting a seasonal pattern

```
>>> sd = sm.tsa.seasonal_decompose(temp_series, freq=12)
>>> sd.plot()
```

trend looks similar to x13 trend

**Comment by josef-pkt** [comment]
AFAICS, the relevant option is `trading=True` I think that compensates for unequal number of days per month. (An alternative might be to rescale the monthly numbers by number of days, so it's daily average of the month.)

(I rescaled numbers by dividing by 10000 which changes the y-scale in the plot)

![x13_seasadj](https://cloud.githubusercontent.com/assets/440735/23781413/24e8c7b8-051b-11e7-865b-609a0197eeaa.png)
selected model with the below options: ARIMA Model:  (0 1 1)(1 1 2)

```
x13results = sm.tsa.x13_arima_analysis(endog = temp_series,x12path=XPATH, 
                                       outlier=True,print_stdout=True, freq='M',
                                       trading=True,
                                       maxorder=(4,2), retspec=True)
```

plot looks similar without maxorder

**Comment by josef-pkt** [comment]
FAQ label given that it seems to be working

**Comment by josef-pkt** [comment]
BTW: a working example is always easier to figure out, but I had to adjust the statsmodels imports to get to x13

**Comment by as4456** [comment]
Hi Josef,
Thanks for the prompt replies. R basically has default holiday regression effects for trading days and other holidays like Easter. If set to NULL, would result in a similar series as generated by Python earlier.
Although capturing calendar effects(Trading days, moving holidays) would be significant in adjusting for seasonality, I wanted to know if seasonal difference argument in maxdiff tuple (seasonal difference set as 1)  is currently handling for seasonal variation (since it would make sense to seasonally adjust for variations during summer and winter) to see a more clearer trend.

But the main issue was to get a similar output and that seems correct now.

Thanks

**Comment by bashtage** [comment]
Closing as answered.
